src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs

```diff
@@ -55,22 +59,73 @@ public override int Read()
 
         private void UpdateState()
         {
-            if (_position < _buffer.Length)
+            if (_cachedLineInfo.Span.Contains(_position))
             {
-                var chr = _buffer.CharAt(_position);
-                _current = chr.Character;
-                _location = chr.Location;
+                _location = new SourceLocation(_sourceDocument.FilePath, _position, _cachedLineInfo.LineIndex, _position - _cachedLineInfo.Span.Start);
+                _current = _sourceDocument[_location.AbsoluteIndex];
+
+                return;
             }
-            else if (_buffer.Length == 0)
+
+            if (_position < _sourceDocument.Length)
             {
-                _current = -1;
-                _location = SourceLocation.Zero;
+                if (_position >= _cachedLineInfo.Span.End)
+                {
+                    // Try to avoid the GetLocation call by checking if the next line contains the position
+                    int nextLineIndex = _cachedLineInfo.LineIndex + 1;
```

Comment:
```suggestion
                    var nextLineIndex = _cachedLineInfo.LineIndex + 1;
```
Similarly in other places as well